src/store/modules/ModDownloadModule.ts

```diff
@@ -0,0 +1,97 @@
+import { ImmutableProfile } from '../../model/Profile';
+import ThunderstoreCombo from '../../model/ThunderstoreCombo';
+import ThunderstoreMod from '../../model/ThunderstoreMod';
+import ThunderstoreVersion from '../../model/ThunderstoreVersion';
+import ThunderstoreDownloaderProvider from '../../providers/ror2/downloading/ThunderstoreDownloaderProvider';
+
+import { State as RootState } from '../index';
+
+interface Progress {
+    progress: number;
+    status: number; // StatusEnum
+}
+
+interface ModProgressItem {
+    mod: ThunderstoreMod;
+    version: ThunderstoreVersion;
+    installProgress: Progress;
+    downloadProgress: Progress;
+}
+
+interface State {
+    allDownloads: ModProgressItem[];
+}
+
+/**
+ * State for Mod Downloadning/Installation.
+ */
+export default {
+    namespaced: true,
+
+    state: (): State => ({
+        allDownloads: [],
+    }),
+
+    getters: <GetterTree<State, RootState>>{
+        allDownloads(state): ModProgressItem[] {
+            return state.allDownloads;
+        },
+        activeDownload(state): ModProgressItem {
+            return state.allDownloads.slice(-1)[0]; // Last element of the array
+        },
+        activeDownloadProgress(state): number | undefined {
+            if(state.allDownloads.length > 0) {
+                return state.allDownloads.slice(-1)[0].downloadProgress.progress; // Last element of the array
+            }
+        },
+    },
+    mutations: {},
+    actions: <ActionTree<State, RootState>>{
+
+        //TODO: Do the installation
+        async downloadAndInstallMod(
+            {dispatch, getters, state},
+            params: {
+                profile: ImmutableProfile,
+                mod: ThunderstoreMod,
+                version: ThunderstoreVersion,
+            }
+        ): Promise<boolean> {
+            state.allDownloads.push({
+                mod: params.mod,
+                version: params.version,
+                installProgress: { progress: 0, status: StatusEnum.PENDING },
+                downloadProgress: { progress: 0, status: StatusEnum.PENDING }
+            });
+
+            return new Promise((resolve, reject) => {
+                ThunderstoreDownloaderProvider.instance.download(
+                    params.profile,
+                    params.mod,
+                    params.version,
+                    true,
+                    (progress, modName, status, err) => {
+                        const index = state.allDownloads.findIndex((modProgressItem) => modProgressItem.mod.getName() === modName);
+
+                        if (status === StatusEnum.FAILURE) {
+                            if(err) {
+                                reject(err);
+                            }
+                        }
+                        if (status === StatusEnum.PENDING) {
+                            if (index !== -1) {
+                                state.allDownloads[index].downloadProgress = { progress: progress, status: status };
```

Comment:
IIRC only mutations should, well, mutate the state, otherwise the change might not propagate correctly or there could be race conditions. So that's why this whole callback should be a mutation. (I know there's one getter is TsModsModule that also does this but it's clearly labeled a bad idea and done because unlike actions, getters don't have access to mutations.